### PR TITLE
Gateway caching proposal

### PR DESCRIPTION
**Pull request by zkokelj**
### Why this change is needed

To discuss if and how to best implement caching on Ten Gateway.

https://github.com/ten-protocol/ten-internal/issues/2851

### What changes were made as part of this PR

Please provide a high level list of the changes made

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: design/gateway/gateway_caching.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+- `eth_accounts`
+- `eth_chainID`
+- `eth_coinbase`
+- `eth_getBlockByHash`
+- `eth_getBlockByNumber`
+- `eth_getBlockRecipients`
+- `eth_getBlockTransactionCountByHash`
+- `eth_getBlockTransactionCountByNumber`
+- `eth_getCode`
+- `eth_getStorageAt (Not used)`
```
Comment (BedrockSquirrel) [inline design/gateway/gateway_caching.md:25]: Technically this one is used, we override it to use it for the custom query stuff. The custom query stuff might be cacheable, it doesn't include block number.

File: design/gateway/gateway_caching.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# Ten Gateway Caching Design
+
+## 1. Why cache requests on Gateway?
+
+Currently, all `eth_` requests that hit the gateway are forwarded to our Sequencer.
```
Comment (tudor-malene) [inline design/gateway/gateway_caching.md:5]: the sequencer is temporary. Let's say to the Ten Nodes and are executed by the Ten enclave

File: design/gateway/gateway_caching.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+- `eth_getTransactionByBlockNumberAndIndex`
+- `eth_getTransactionByHash`
+- `eth_getTransactionReceipt`
+- `eth_getUncleCountByBlockHash`
+- `eth_getUncleCountByBlockNumber`
+- `eth_maxPriorityFeePerGas`
+- `eth_sign`
+- `eth_signTransaction`
+
+Non-cacheable request methods are:
```
Comment (tudor-malene) [inline design/gateway/gateway_caching.md:36]: they are also cache-able, but with a TTL until the next batch.

**Review by tudor-malene — COMMENTED** [review]
I think all calls are cache-able.
The TTL differs.

Also, would be good to touch for "eth_call" - for example, what would be the key for caching

**Review by badgersrus — COMMENTED** [review]
Given we're expecting _all_ user traffic to come through this gateway, we might want to consider a standalone cache for testnet/ mainnet. This would reduce the memory overhead but you'd just get a bit of additional infrastructure complexity. We can add specific rate-limiting for the non-cacheable methods to prevent a targeted DDoS via those. 

If this is something we'd want to consider you would want to define a cache interface with an in-memory implementation for dev and a redis/ memcached implementation that we could use for production which would be specified in the application configuration.

**Comment by BedrockSquirrel** [comment]
It would be nice to have some indication of the value of a cache before investing too much time and resources. Intuitively, it feels to me like we're going to have a tiny cache hit ratio overall (like I'd guess < 1%).

Maybe I'm wrong, or might be that even the savings from occasional cache hits for certain methods make it worthwhile. But I'd be interested to hear any hunches or see some back of an envelope maths or finger in the air stats to motivate a cache.

Feels a bit like premature optimisation to me.

**Comment by badgersrus** [comment]
> It would be nice to have some indication of the value of a cache before investing too much time and resources. Intuitively, it feels to me like we're going to have a tiny cache hit ratio overall (like I'd guess < 1%).
> 
> Maybe I'm wrong, or might be that even the savings from occasional cache hits for certain methods make it worthwhile. But I'd be interested to hear any hunches or see some back-of-an-envelope maths or finger-in-the-air stats to motivate a cache.
> 
> Feels a bit like premature optimisation to me.

Are there any prometheus/ grafana dashboards we can use to track the requests being made? That way we could quantify the % of hits and extrapolate out to anticipated traffic

**Comment by anixon604** [comment]
> > It would be nice to have some indication of the value of a cache before investing too much time and resources. Intuitively, it feels to me like we're going to have a tiny cache hit ratio overall (like I'd guess < 1%).
> > Maybe I'm wrong, or might be that even the savings from occasional cache hits for certain methods make it worthwhile. But I'd be interested to hear any hunches or see some back-of-an-envelope maths or finger-in-the-air stats to motivate a cache.
> > Feels a bit like premature optimisation to me.
> 
> Are there any prometheus/ grafana dashboards we can use to track the requests being made? That way we could quantify the % of hits and extrapolate out to anticipated traffic

I think it might be viable to just give a go on a OOB caching package before building a custom one out. Just to see what sane defaults might accomplish and get quick functionality. 

Like this one https://github.com/dgraph-io/ristretto will even export metrics it seems. So it can just bolt on and run for a bit and see how the metrics play out.

@zkokelj @badgersrus

File: design/gateway/gateway_caching.md (hunk 3 above)
Comment (zkokelj) [reply design/gateway/gateway_caching.md:36]: Fixed. What do you think about `eth_sendRawTransaction` we can probably also catch with TTL until next batch.
(Assuming the key is the whole transaction and submitting the same tx twice in short time period should give the same response).

**Comment by zkokelj** [comment]
Here is stats from Sepolia Ten Gateway in the past ~6 days.

```
eth_accounts: 0
eth_chainID: 0
eth_coinbase: 0
eth_getBlockByHash: 17241
eth_getBlockByNumber: 500670
eth_getBlockRecipients: 0
eth_getBlockTransactionCountByHash: 0
eth_getBlockTransactionCountByNumber: 0
eth_getCode: 22086
eth_getStorageAt: 49096
eth_getTransactionByBlockHashAndIndex: 0
eth_getTransactionByBlockNumberAndIndex: 0
eth_getTransactionByHash: 2913
eth_getTransactionReceipt: 454676
eth_getUncleCountByBlockHash: 0
eth_getUncleCountByBlockNumber: 0
eth_maxPriorityFeePerGas: 0
eth_sign: 0
eth_signTransaction: 0
eth_getBalance: 561268
eth_blockNumber: 1760227
eth_call: 402467
eth_createAccessList: 0
eth_estimateGas: 24998
eth_feeHistory: 3070
eth_getProof: 0
eth_gasPrice: 280329
eth_getFilterChanges: 0
eth_getFilterLogs: 0
eth_getTransactionCount: 114349
eth_newBlockFilter: 0
eth_newFilter: 0
eth_newPendingTransactionFilter: 0
eth_sendRawTransaction: 88151
eth_sendTransaction: 0
eth_syncing: 0
eth_uninstallFilter: 0
```

There are a few methods that get called a lot, more than half of the methods were never called in the past week. 
I think it might be worth implementing a simple cache for those. `eth_getBlockByNumber` looks like a method that can be cached the best  (and also for longer time) + some other can use shorter TTL

**Comment by zkokelj** [comment]
> I think all calls are cache-able. The TTL differs.
> 
> Also, would be good to touch for "eth_call" - for example, what would be the key for caching

I looked into `eth_call` and for the key we can use combination of: `from`,`to`, `data`, `block`. Because block can be also `latest` we should not have very long TTL for that cache.

**Comment by otherview** [comment]
Good stuff, and an important proposal.

- I believe that the caching is different for the below as they can use `latest, earliest, pending or blockNumber`.
The following methods have an extra default block parameter:
. eth_getBalance
. eth_getCode
. eth_getTransactionCount
. eth_getStorageAt
. eth_call



- It's likely that caching can be done by a different app rather than inMem
The typical eth response is a few kb, I think that a first iteration done in Memory is good but might fill up fast. Also it's likely that it will take some effort to easily scale or be as resilient as perhaps a solution like Redis .

**Review by moraygrieve — APPROVED** [review]
LGTM

**Pull request merged**